pkg/agent/agent.go

```diff
@@ -417,3 +387,248 @@ func (f *Flows) onInterfaceAdded(iface ifaces.Interface) {
+			return
+		case event := <-ifaceEvents:
+			slog.WithField("event", event).Debug("received event")
+			switch event.Type {
+			case ifaces.EventAdded:
+				eventAdded(event.Interface)
+			case ifaces.EventDeleted:
+				// qdiscs, ingress and egress filters are automatically deleted so we don't need to
+				// specifically detach them from the ebpfFetcher
+			default:
+				slog.WithField("event", event).Warn("unknown event type")
+			}
+		}
+	}
+}
+
+// ebpfPacketFetcher abstracts the interface of ebpf.FlowFetcher to allow dependency injection in tests
+type ebpfPacketFetcher interface {
+	io.Closer
+	Register(iface ifaces.Interface) error
+
+	LookupAndDeleteMap() map[int][]*byte
+	ReadPerf() (perf.Record, error)
+}
+
+// PacketsAgent instantiates a new agent, given a configuration.
+func PacketsAgent(cfg *Config) (*Packets, error) {
+	plog.Info("initializing Packets agent")
+
+	// configure informer for new interfaces
+	informer := configureInformer(cfg, plog)
+
+	plog.Info("[PCA]acquiring Agent IP")
+	agentIP, err := fetchAgentIP(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("acquiring Agent IP: %w", err)
+	}
+
+	// configure selected exporter
+	packetexportFunc, err := buildPacketExporter(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	ingress, egress := flowDirections(cfg)
+
+	fetcher, err := ebpf.NewPacketFetcher(cfg.CacheMaxFlows, cfg.PCAFilters, ingress, egress)
+	if err != nil {
+		return nil, err
+	}
+
+	return packetsAgent(cfg, informer, fetcher, packetexportFunc, agentIP)
+}
+
+// packetssAgent is a private constructor with injectable dependencies, usable for tests
+func packetsAgent(cfg *Config,
+	informer ifaces.Informer,
+	fetcher ebpfPacketFetcher,
+	packetexporter node.TerminalFunc[[]*flow.PacketRecord],
+	agentIP net.IP,
+) (*Packets, error) {
+	// configure allow/deny interfaces filter
+	filter, err := initInterfaceFilter(cfg.Interfaces, cfg.ExcludeInterfaces)
+	if err != nil {
+		return nil, fmt.Errorf("configuring interface filters: %w", err)
+	}
+
+	registerer := ifaces.NewRegisterer(informer, cfg.BuffersLength)
+
+	interfaceNamer := func(ifIndex int) string {
+		iface, ok := registerer.IfaceNameForIndex(ifIndex)
+		if !ok {
+			return "unknown"
+		}
+		return iface
+	}
+
+	perfTracer := flow.NewPerfTracer(fetcher, cfg.CacheActiveTimeout)
+
+	packetbuffer := flow.NewPerfBuffer(cfg.CacheMaxFlows, cfg.CacheActiveTimeout)
+
+	return &Packets{
+		ebpf:           fetcher,
+		interfaces:     registerer,
+		filter:         filter,
+		cfg:            cfg,
+		packetbuffer:   packetbuffer,
+		perfTracer:     perfTracer,
+		agentIP:        agentIP,
+		interfaceNamer: interfaceNamer,
+		exporter:       packetexporter,
+	}, nil
+}
+
+func buildPacketExporter(cfg *Config) (node.TerminalFunc[[]*flow.PacketRecord], error) {
+	if cfg.TargetPort == 0 {
+		return nil, fmt.Errorf("missing target port: %d",
+			cfg.TargetPort)
+	}
+	pcapStreamer, err := exporter.StartPCAPSend(fmt.Sprintf("%d", cfg.TargetPort))
```

Comment:
I feel it is misleading to reuse the `cfg.TargetPort` here as this is used for the agent _listening_ as a server, unlike its usage for flows where it is use for sending (as a client). I guess that's why I was misguided the other day and I was trying to configure something server-side to receive packets.

I would suggest either to create new parameters to clarify that the agent will be _listening_ on a port, or to reverse the approach and have the agent send the pcap to a receiver. @shach33  @msherif1234 @dushyantbehl wdyt is better?